app/Main.hs

```diff
@@ -14,24 +15,38 @@ import Hie.Locate
 import Hie.Yaml
 import System.Directory
 import System.Directory.Internal
-import System.FilePath.Posix
+import System.FilePath
+import Hie.CabalHelper
+import Distribution.Helper (Ex (..))
 
 main :: IO ()
 main = do
   pwd <- getCurrentDirectory
-  files <- listDirectory pwd
-  let name =
-        if  | any (("dist-newstyle" ==) . takeFileName) files -> "cabal"
-            | any ((".stack-work" ==) . takeFileName) files -> "stack"
-            | any (("cabal.project" ==) . takeFileName) files -> "cabal"
-            | any (("stack.yaml" ==) . takeFileName) files -> "stack"
-            | otherwise -> "cabal"
-  cfs <- runMaybeT $ case name of
-    "cabal" -> cabalPkgs pwd
-    _ -> stackYamlPkgs pwd
-  when (null cfs) $ error $
-    "No .cabal files found under"
-      <> pwd
-      <> "\n You may need to run stack build."
-  pkgs <- catMaybes <$> mapM (nestedPkg pwd) (concat cfs)
-  putStr <$> hieYaml name $ fmtPkgs name pkgs
+  projM <- findCabalHelperEntryPoint $ pwd </> "Foo.hs"
+  case projM of
+    Nothing -> error "could not find project context"
+    Just proj@(Ex p) -> do
+      env <- initialiseEnvironment p
+      packages <- loadPackages env
+      pkgs <- toHieYaml env packages
+      let name = if | isCabalProject proj -> "cabal"
```

Comment:
How does the logic of `isCabalProject` compare to `| any (("dist-newstyle" ==)`?

Reply:
That is something to worry about for later. In this case, the project is decided based on the found files, e.g. `cabal.project`, `stack.yaml` and `*.cabal` files. In this order, a project type is selected.

Reply:
At some point, we should take build directory into account first so the cabal-helper implicit logic marches the default logic and builds with the users preferred tool.